integration-tests/tester/framework/node.go

```diff
@@ -0,0 +1,148 @@
+	// after container is started we can get its IP
+	ip, err := dockerContainer.IP(network.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	iotaAPI, err := api.ComposeAPI(api.HTTPClientSettings{URI: getWebAPIBaseURL(ip)})
+	if err != nil {
+		return nil, fmt.Errorf("can't instantiate Web API: %w", err)
+	}
+
+	testWebAPI := NewWebAPI(getWebAPIBaseURL(ip))
+
+	return &Node{
+		Name: name,
+		Profiler: Profiler{
+			pprofURI:     fmt.Sprintf("http://%s:6060", ip),
+			websocketURI: fmt.Sprintf("ws://%s:8081/ws", ip),
+			targetName:   name,
+			Client: http.Client{
+				Timeout: 2 * time.Minute,
+			},
+		},
+		IP:              ip,
+		Config:          cfg,
+		Identity:        identity,
+		WebAPI:          iotaAPI,
+		DebugWebAPI:     testWebAPI,
+		DockerContainer: dockerContainer,
+	}, nil
+}
+
+func (p *Node) String() string {
+	return fmt.Sprintf("Node:{%s, %s, %s, %d}", p.Name, p.ID().String(), p.APIURI(), p.TotalNeighbors())
+}
+
+// APIURI returns the URL under which this node's web API is accessible.
+func (p *Node) APIURI() string {
+	return getWebAPIBaseURL(p.Name)
+}
+
+// TotalNeighbors returns the total number of neighbors the peer has.
+func (p *Node) TotalNeighbors() int {
+	return len(p.neighbors)
+}
+
+// SetNeighbors sets the neighbors of the peer accordingly.
+func (p *Node) SetNeighbors(peers []*peer.Info) {
+	p.neighbors = make([]*peer.Info, len(peers))
+	copy(p.neighbors, peers)
+}
+
+// Spam spams zero value transactions on the node.
+// Returns the number of spammed transactions.
+func (p *Node) Spam(dur time.Duration, depth int, parallelism int, batchSize ...int) (int32, error) {
+	targetAddr, err := checksum.AddChecksum(consts.NullHashTrytes, true, consts.AddressChecksumTrytesSize)
+	if err != nil {
+		return 0, err
+	}
+	spamTransfer := []bundle.Transfer{{Address: targetAddr}}
+	bndl, err := p.WebAPI.PrepareTransfers(consts.NullHashTrytes, spamTransfer, api.PrepareTransfersOptions{})
+	if err != nil {
+		return 0, err
+	}
+
+	batch := 20
+	if len(batchSize) > 0 {
+		batch = batchSize[0]
+	}
+
+	s := time.Now()
+	end := s.Add(dur)
+	var spammed int32
+	var wg sync.WaitGroup
+	wg.Add(parallelism)
+	for j := 0; j < parallelism; j++ {
+		time.Sleep(250 * time.Millisecond)
+		go func() {
+			defer wg.Done()
+			for time.Now().Before(end) {
+
+				// we use batching for increased throughput
+				toSend := make([]trinary.Trytes, batch)
+				for i := 0; i < batch; i++ {
+					tips, err := p.WebAPI.GetTransactionsToApprove(uint64(depth))
+					if err != nil {
+						return
+					}
+					readyTx, err := p.WebAPI.AttachToTangle(tips.TrunkTransaction, tips.BranchTransaction, uint64(p.Config.Coordinator.MWM), bndl)
+					if err != nil {
+						return
+					}
+					toSend[i] = readyTx[0]
+				}
+
+				if _, err = p.WebAPI.BroadcastTransactions(toSend...); err != nil {
+					return
+				}
+
+				atomic.AddInt32(&spammed, int32(batch))
+			}
+		}()
+	}
+	return spammed, nil
```

Comment:
Shouldn't you wait for the waitgroup to be done?

Reply:
Good catch.